### ISSUE TITLE
Certain tensor dynamic indexing not supported by onnx converter? A future feature or a bug?

### ISSUE THREAD

**Issue by yjhe** [onnx/onnx#3510]
### Question
I was using transformers code base with customized model structures, specifically, model's forward() function takes in 2 index tensors (mention_idx and entity_idx, mention_idx of dimension batch_size x 2, entity_idx is of dimension batch_size x 1), the index tensor will need to operate on sequence_output which is of dimension batch_size x sequence_len x hidden_layer_size, to keep only the token embeddings at the corresponding index position. The index position differs across samples, thus the dummy sample indexer idx0 is introduced. Below is the code snippet of above:

      batch_size = sequence_output.shape[0]
      idx0 = [[x] for x in range(batch_size)]
      mention_embeds = sequence_output[idx0, mention_idx, :]
      mention_embeds = torch.mean(mention_embeds, dim=1)
      entity_embeds = torch.squeeze(sequence_output[idx0, entity_idx, :], 1)
      sequence_output = torch.cat([mention_embeds, entity_embeds], dim=1)

Below is what I observed:
1. ONNX conversion (torch.onnx.export()) of above with **single input** (i.e. batch_size =1 ), dynamic axes and onnx checker enabled, didn't complain any issues. However when using the resulting model on batch inferencing (batch_size > 1), only the first sample gives the correct prediction.
2. ONNX conversion (torch.onnx.export()) of above with **batch input** (e.g. batch_size =5 ), dynamic axes and onnx checker enabled, didn't complain any issues. However when using the resulting model on batch inferencing, only when batch_size = 5 gives correct results for all samples in batch, all other batch size (which I expected dynamic axes to work) will report error saying new batch_size can't be broadcast with the conversion batch_size 5:

> onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running Add node. Name:'Add_354' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/math/element_wise_ops.h:479 void onnxruntime::BroadcastIterator::Init(ptrdiff_t, ptrdiff_t) axis == 1 || axis == largest was false. Attempting to broadcast an axis by a dimension other than 1. 3 by 5

I felt the issue lies in the above code snippet that ONNX is not able to dynamically indexing. So I changed from idx0 to simple for-loop which dynamically decides batch size based on input, and accumulate resulting embeddings in a new tensor, now the batch_size used in conversion is fixed for all future inference, despite use of dynamic axes. Onnx model inference results are correct, however e.g. if the input used for export() has batch_size k, onnx model inference output will always be size k. If inference input is of size m, and k < m, inference output will generate only k results (all correct), if k > m, inference output first m results are correct, the rest seems to be garbage values.

It seems the model conversion is not compatible with the usage of dynamic indexing, please advise on how to resolve.

Below is the environment on 4.15.0-140-generic 144-Ubuntu SMP OS:
- onnx                    1.9.0
- onnxruntime-gpu         1.7.0
- onnxruntime-tools       1.7.0
- torch                   1.8.0+cu111

**Comment by jcwchen** [comment]
Hi @yjhe,
It seems like a question for the converter. Please open an issue under pytorch repo to get the best help from the converter team (torch.onnx.export). Thanks.
